kube-runtime/src/reflector/store.rs

```diff
@@ -96,48 +99,75 @@ where
             }
-            watcher::Event::Deleted(obj) => {
+            watcher::Event::Delete(obj) => {
                 let key = obj.to_object_ref(self.dyntype.clone());
                 self.store.write().remove(&key);
             }
-            watcher::Event::Restarted(new_objs) => {
+            watcher::Event::RestartInit => {
+                self.buffer = AHashMap::new();
+            }
+            watcher::Event::RestartPage(new_objs) => {
                 let new_objs = new_objs
                     .iter()
                     .map(|obj| (obj.to_object_ref(self.dyntype.clone()), Arc::new(obj.clone())))
                     .collect::<AHashMap<_, _>>();
-                *self.store.write() = new_objs;
+                self.buffer.extend(new_objs);
             }
-        }
+            watcher::Event::Restart => {
+                let mut store = self.store.write();
+
+                // Swap the buffer into the store
+                std::mem::swap(&mut *store, &mut self.buffer);
 
-        // Mark as ready after the first event, "releasing" any calls to Store::wait_until_ready()
-        if let Some(ready_tx) = self.ready_tx.take() {
-            ready_tx.init(())
+                // Clear the buffer
+                // This is preferred over self.buffer.clear(), as clear() will keep the allocated memory for reuse.
+                // This way, the old buffer is dropped.
+                self.buffer = AHashMap::new();
+
+                // Mark as ready after the RestartedDone, "releasing" any calls to Store::wait_until_ready()
+                if let Some(ready_tx) = self.ready_tx.take() {
+                    ready_tx.init(())
+                }
+            }
+            watcher::Event::RestartApply(obj) => {
+                let key = obj.to_object_ref(self.dyntype.clone());
+                let obj = Arc::new(obj.clone());
+                self.buffer.insert(key, obj);
+            }
+            watcher::Event::RestartDelete(obj) => {
+                let key = obj.to_object_ref(self.dyntype.clone());
+                self.buffer.remove(&key);
+            }
         }
     }
 
     /// Broadcast an event to any downstream listeners subscribed on the store
     pub(crate) async fn dispatch_event(&mut self, event: &watcher::Event<K>) {
         if let Some(ref mut dispatcher) = self.dispatcher {
             match event {
-                watcher::Event::Applied(obj) => {
+                watcher::Event::Apply(obj) | watcher::Event::RestartApply(obj) => {
                     let obj_ref = obj.to_object_ref(self.dyntype.clone());
                     // TODO (matei): should this take a timeout to log when backpressure has
                     // been applied for too long, e.g. 10s
                     dispatcher.broadcast(obj_ref).await;
                 }
 
-                watcher::Event::Restarted(new_objs) => {
+                watcher::Event::RestartPage(new_objs) => {
```

Comment:
Hm, so the way subscribers are set up atm:
- Once an event is received, the object is stored, and then an object reference is broadcast to any subscribers.
- Each subscriber will pull out the object by doing a read from the store. Each subscriber has dependency on a store (at the moment).

I'm not sure sending pages to subscribers will work well, since my understanding is that the store is atomically updated once we've buffered all resynced objects. Without having a handle to the buffer, subscribed readers won't be able to read anything.

I think we have a couple of options here, but it's a lot of spitballing on my part:
- We could have the dispatcher send a refresh of the entire store once it observes a `Restart` event -- that signals downstream readers they can now start to act on the new state.
- We could let subscribers hold a handle to the buffer and let them handle pages.
- We could maybe change dispatch to send objects instead of references, we could probably work with that.

Let me know if any of the solution above make sense. We can also put a pin in this and try to get it to work after this PR. Happy to chat through things either here or on discord!

Reply:
Applied the latest solution proposed in https://github.com/kube-rs/kube/pull/1494/commits/1cc5b898a41f9432296dd6534b7bf8ee1d6d6b7f